src/utilities/vaccineReport.js

```diff
@@ -0,0 +1,238 @@
+    [BREACH_SECTION_FIELDS.TYPE]: CONFIG_TYPE_TO_NAME[breach.type],
+    [BREACH_SECTION_FIELDS.START]: moment(breach.startTimestamp).format('DD/MM/YYYY hh:mm:ss a'),
+    [BREACH_SECTION_FIELDS.END]: breach.endTimestamp
+      ? moment(breach.endTimestamp).format('DD/MM/YYYY hh:mm:ss a')
+      : '',
+    [BREACH_SECTION_FIELDS.DURATION]: `${breach.duration.humanize()} ${
+      breach.endTimestamp ? '' : '(ongoing)'
+    }`,
+    [BREACH_SECTION_FIELDS.MIN_TEMPERATURE]: temperature(breach.minimumTemperature).format(),
+    [BREACH_SECTION_FIELDS.MAX_TEMPERATURE]: temperature(breach.maximumTemperature).format(),
+  }));
+
+  const title = SECTION_TITLES.BREACHES;
+
+  return { title, data, parser };
+};
+
+// TODO: LOCALIZE
+const createLogSection = logs => {
+  const parser = getParser(LOG_SECTION_FIELDS);
+
+  const title = SECTION_TITLES.LOGS;
+  const data = logs.map(log => ({
+    [LOG_SECTION_FIELDS.TIMESTAMP]: moment(log.timestamp).format('DD/MM/YYYY hh:mm:ss a'),
+    [LOG_SECTION_FIELDS.TEMPERATURE]: temperature(log.temperature).format(),
+    [LOG_SECTION_FIELDS.BREACH]: log.breach ? 'Yes' : 'No',
+    [LOG_SECTION_FIELDS.LOG_INTERVAL]: Math.round(log.logInterval / SECONDS.ONE_MINUTE),
+  }));
+
+  return { title, parser, data };
+};
+
+function Report() {
+  this.sections = [];
+
+  this.addSection = section => {
+    this.sections.push(section);
+  };
+
+  this.build = options => {
+    const { sectionBreak } = options;
+
+    return this.sections
+      .map(
+        ({ title = '', parser, data }) =>
+          `${title}\n${parser.parse(data)}${sectionBreak ? '\n\n' : ''}`
+      )
+      .join('');
+  };
+}
+
+export const vaccineReport = async (sensor, user, comment) => {
+  const report = new Report();
+
+  const generalSection = await createGeneralSection(sensor, user, comment);
+  const sensorSection = createLoggingSection(sensor);
+  const sensorStatsSection = createSensorStatsSection(sensor);
+  const breachConfigsSection = createBreachConfigSection(sensor?.breachConfigs);
+  const breachSection = createBreachSection(sensor?.breaches.sorted('startTimestamp', false));
+  const logSection = createLogSection(sensor?.logs.sorted('timestamp', false));
+
+  report.addSection(generalSection);
+  report.addSection(sensorSection);
+  report.addSection(sensorStatsSection);
+  report.addSection(breachConfigsSection);
+  report.addSection(breachSection);
+  report.addSection(logSection);
+
+  const content = report.build({ sectionBreak: true });
+
+  return content;
+};
+
+const writeReport = async content => {
+  const directory = '/Download/mSupplyMobile_data';
+  const now = moment().format('DD-MM-YYYY-HHmm');
+  const file = `/${now}.csv`;
+  const path = `${RNFS.ExternalStorageDirectoryPath}${directory}${file}`;
+
+  try {
+    await RNFS.mkdir(`${RNFS.ExternalStorageDirectoryPath}${directory}`);
+    await RNFS.writeFile(path, content, 'utf8');
+    // eslint-disable-next-line no-empty
+  } catch (e) {}
```

Comment:
maybe Toast on error? And don't try to email if the file cannot be written.

Reply:
```EACCES (Permission denied), open '/storage/emulated/0/Download/mSupplyMobile_data/04-02-2021-1115.csv'```

Reply:
Ah! Of course, totally forgot about permissions 🙄 Another one that I forgot to implement was a check for if there was any temperature logs - getting an empty report is a bit average!

I just spent a bit thinking of the best way to do this and checking out the `UIDatabase` export data thingy. I think it might be best to do the checks before this is called, like export data. Keeps the toast out of these functions which have nothing to do with UI stuff and also allows for easier use of checking the permission in the redux store and dispatching an action if the permission is updated etc.